pkg/controller/useraccount/useraccount_controller.go

```diff
@@ -107,16 +112,28 @@ func (r *ReconcileUserAccount) Reconcile(request reconcile.Request) (reconcile.R
 		return reconcile.Result{}, err
 	}
 
-	var createdOrUpdated bool
-	var user *userv1.User
-	if user, createdOrUpdated, err = r.ensureUser(reqLogger, userAcc); err != nil || createdOrUpdated {
-		return reconcile.Result{}, err
-	}
+	// If the UserAccount has not been deleted, create or update user and identity resources.
+	// If the UserAccount has been deleted, delete secondary resources identity and user.
+	if !util.IsBeingDeleted(userAcc) {
+		// Add the finalizer if it is not present
+		if err := r.addFinalizer(userAcc, userAccFinalizerName); err != nil {
+			return reconcile.Result{}, err
+		}
 
-	if _, createdOrUpdated, err = r.ensureIdentity(reqLogger, userAcc, user); err != nil || createdOrUpdated {
-		return reconcile.Result{}, err
-	}
+		var createdOrUpdated bool
+		var user *userv1.User
+		if user, createdOrUpdated, err = r.ensureUser(reqLogger, userAcc); err != nil || createdOrUpdated {
+			return reconcile.Result{}, err
+		}
 
+		if _, createdOrUpdated, err = r.ensureIdentity(reqLogger, userAcc, user); err != nil || createdOrUpdated {
+			return reconcile.Result{}, err
+		}
+	} else if util.HasFinalizer(userAcc, userAccFinalizerName) {
+		if err = r.manageCleanUpLogic(reqLogger, userAcc); err != nil {
```

Comment:
Can we rename `manageCleanUpLogic` ?

Reply:
@dipak-pawar this name is from the library as the method to implement. Not sure we should renmae this if we are going the library approach?

Reply:
@tinakurian, my point here is if we can name a function from which we can understand what's happening inside the function. e.g.`cleanUpUserAndIdentiy()`, then it'd be good.